{% extends "layout.html" %}
{% block body %}

    <div class="page-header" style="margin-top: 12px; border-bottom-color: #fff;"></div>

    <div class="panel">
        <div class="panel-heading panel-title"><h3 style="color: #f0ad4e;">Documentation</h3></div>
{#        <div class="panel-body" >#}
{#            <p class="lead">Detailed documentation coming soon!</p>#}
{#        </div>#}
    </div>

    <div class="panel">
        <div class="panel-heading panel-title"><h3 class="panel-title">Brief Overview</h3></div>
        <div class="panel-body" >
            <p class="lead">Heterologous protein expression is an important method in proving homologous functions of proteins, in the synthetic construction of genetic circuits and in overexpressing proteins for biopharmaceutical applications and structural biology research, to name a few. These research efforts have highly profited from the exponentially increasing sequence information from genome and metagenome projects and the considerable decrease in costs and increase in speed for synthetic genes providing a convenient route to obtain genes encoding these virtual proteins.</p>
            <p class="lead" style="text-align: justify;">There are global and local factors that influence the expression level of genes in heterologous hosts. The most important global factors are the different GC contents in coding genes between species and the different codon context, which specifies the non-random occurrence of codon pairs. Local variables include mRNA secondary structures close to translation initiation sites, internal repeats and patterns resembling ribosomal binding sites. Available software for gene design use the most commonly used synonymous codons for the target gene and allow to exclude some of the local restraints.</p>
            <p class="lead" style="text-align: justify;">Here, we present a software for the design of typical sequences for heterologous protein expression, that analyses available structural and genomic data in terms of sequences introducing ribosomal pausing. We think that typical genes should be better for high-level recombinant expression avoiding many of the problems caused by the overuse of a small number of tRNAs. This includes introducing ribosomal pausing regions that have been shown to be essential for the correct folding of proteins in many cases.</p>
        </div>
    </div>

{% endblock %}
